README.md
# elm-dependencies-analyzer

A little program I created for analyzing package dependencies of application elm.json.

This is my first Elm program of over 1000 lines, and code quality clearly reflects that. :D

Live version: https://www.markuslaire.com/github/elm-dependencies-analyzer/

## Usage

### Loading elm.json

1) Open your application `elm.json` or copy-paste it into the editor.
   - You can also load an example with some conflicting packages with "Load example".
2) Press "Analyze" to parse JSON and load package information.
   - Loaded information is cached for 24 hours in browser and in backend cache.
     If cache has become stale, loading can take a moment while cache is being refreshed, but subsequent loads will go faster.

Once `elm.json` has been analyzed, table on right will show all relevant packages.
For each package the version in `elm.json` and all newer versions will be shown.

### Packages

- Packages are shown in four sections based on `elm.json`:
  direct, indirect, test-direct and test-indirect
- Direct packages have checkmark and black font color.
- Indirect packages have gray font color.
- Use checkboxes to change whether package is considered direct package or not.
- Test-dependencies are marked with `TEST`.
- Packages which are mentioned in `elm.json`,
  but not needed by current direct packages, are marked with strikethrough.
- Additional packages which are not mentioned in `elm.json`,
  are marked with `NEW` and listed at end of each section.

### Selected versions

- Selected version of each package is shown with blue background. To change selected version, click on any other version.
- If there are any dependency conflicts with selected versions, those are listed under the table.
- If there are no conflicts with selected versions,
  then any non-selected version which is incompatible with selected versions, will be shown with red background.

### Incompatible versions

- When hovering cursor over any package version,
  all versions which are incompatible with that one will be marked with red border.

## Limitations

- Only application `elm.json` can be analyzed, not package `elm.json`.
  - If you would find it useful to also support package `elm.json`, please open an issue.
- Moving packages between normal/test dependencies is not fully supported yet.

## Building

1) [Install Elm 0.19](https://guide.elm-lang.org/install.html) 
2) Run `elm make src/*.elm --optimize --output=elm.js` in project base directory.
   - Optionally you can also run minifier like `uglifyjs`.
     See included `make.sh` for some building options.
     (This is just a generic script for running `elm make`, not specific to this project.)
3) Open `index.html` in browser.

### Backend cache

- `cacheUrl` in `Backend.elm` sets the backend cache to be used.
- My cache at `https://www.markuslaire.com/github/elm-dependencies-analyzer/backend/cache.php?` is available
  publicly for now. I intend to keep this cache public, but if I start getting too much traffic,
  I could make it private.
- The backend script is included in this repository as `cache.php`, so you could also just run your own cache.

### Backend error reporting

- `logErrorUrl` in `Backend.elm` enables/disables automatic error reporting,
  and sets the URL to be used when enabled.
- For custom builds this should be set to either `Nothing` (disabled) or to your own URL.
   - (Error reports from custom builds aren't that useful to me,
     as I can't know what has been changed in code.)


CHANGELOG.md
# Changelog

Only a selection of changes are listed here.

## 2018-12-14
- Automatic reporting of (some) internal errors

## 2018-12-09
- Include test-dependencies
  - Packages are now shown in four sections:
    direct, indirect, test-direct & test-indirect
- Create and show new `elm.json`, with download option.

## 2018-12-06
- Packages are now shown in two sections: direct & indirect
  - Section is based on parsed `elm.json`.
  - New packages are shown at end of each section, marked with "NEW".
- New packages can also be marked direct.
- FIXED: Showing red background was ignoring selected versions of
  indirect packages.

## 2018-12-03
- Better reporting of which packages are in conflict.
  - Before: `1.0.1 <= v < 2.0.0 needed by elm/bytes 2.0.0 (and 2 others)`
    - most relevant package `elm/http` hidden behind `(and 2 others)`
  - After: `1.0.1 <= v < 2.0.0 needed by elm/http 2.0.0`
    - only the relevant package is reported


make.sh
#!/bin/bash

#
# This script is based on https://guide.elm-lang.org/optimization/asset_size.html
#

set -e

SOURCES=src/*.elm
JS_TEMP=elm.tmp.js
JS_TARGET=elm.js


if [ "$1" = "" ]
then
    elm make $SOURCES --optimize --output=$JS_TEMP

    uglifyjs $JS_TEMP --compress \
      'pure_funcs="F2,F3,F4,F5,F6,F7,F8,F9,A2,A3,A4,A5,A6,A7,A8,A9",pure_getters,keep_fargs=false,unsafe_comps,unsafe' \
      2>/dev/null | uglifyjs --mangle --output=$JS_TARGET

    echo
    echo "Compiled size: $(cat $JS_TEMP             | wc -c) bytes"
    echo "Minified size: $(cat $JS_TARGET           | wc -c) bytes"
    echo "Gzipped  size: $(cat $JS_TARGET | gzip -c | wc -c) bytes"

    rm $JS_TEMP

elif [ "$1" = "d" ]
then
    elm make $SOURCES --debug --output=$JS_TARGET

elif [ "$1" = "q" ]
then
    elm make $SOURCES --optimize --output=$JS_TARGET

elif [ "$1" = "qq" ]
then
    elm make $SOURCES --output=$JS_TARGET

else
    echo
    echo "USAGE"
    echo "  ./make.sh [d | q | qq]"
    echo
    echo "OPTIONS"
    echo "  d   debug         ; with --debug"
    echo "  q   quick         ; no minifying"
    echo "  qq  really quick  ; no minifying or --optimize"
    echo
    echo
    echo
fi


util/newTag.php
#!/usr/bin/env php
<?php

/*
 * Get 4-digit rangom number which is not yet seen in code.
 */

# ======================================================================
# CONSTANTS

const GLOB_SOURCE_FILES = "src/*.elm";

# ======================================================================
# MAIN

$seen = [];

foreach (glob(GLOB_SOURCE_FILES) as $file) {
    $data = file_get_contents($file);
    if ($data === false) throw new Exception();

    if (preg_match_all("/(\d{4,})/", $data, $m) === false) throw new Exception();
    foreach ($m[1] as $number) {
        for ($n = 0; $n <= strlen($number) - 4; $n++) {
            $seen[substr($number, $n, 4)] = true;            
        }
    }
}

$new = null;
do {
    $new = rand(1000, 9999);
} while (array_key_exists($new, $seen));

printf("Seen %d. New: %d\n", count($seen), $new);


backend/logerror.php
<?php
require_once dirname(__FILE__) . '/config.php';

# ================================================================================
# CONSTANTS

const DIR_LOG_ERROR = config\DIR_LOG . '/error';

# ================================================================================
# MAIN


try {
    $method = $_SERVER['REQUEST_METHOD'];

    if ($method === "OPTIONS") {
        header('Access-Control-Allow-Origin: *');
        header('Access-Control-Allow-Headers: Content-Type');
        header('Access-Control-Allow-Methods: POST, OPTIONS');

    } elseif ($method === "POST") {
        $data = file_get_contents('php://input');

        # LOGFILE
        if ($data !== false) {
            $postJson = json_decode($data, true);
            if ($postJson !== null) {
                if (array_key_exists('tag', $postJson)) {
                    if (preg_match("/^([0-9]{4})$/", $postJson['tag'], $m) === 1) {
                        $logFile = $m[1];
                    } else {
                        $logFile = 'invalidtag';
                    }
                } else {
                    $logFile = 'notag';
                }
            } else {
                $logFile = 'nojson';
            }
        } else {
            $logFile = 'nodata';
            $data    = 'nodata';
        }

        # LOGDATA
        list ($usecAsSec, $sec) = explode(' ', microtime());
        $usec    = intval($usecAsSec * 1000 * 1000);
        $logData = sprintf("%s\nTIME %s.%06s\nORIGIN %s\n%s\n",
            "======================================================================",
            gmdate('Y-m-d H:i:s', $sec),
            $usec,
            $_SERVER['HTTP_ORIGIN'],
            $data
        );

        # MKDIR
        if (! file_exists(DIR_LOG_ERROR)) {
            if (! mkdir(DIR_LOG_ERROR, 0777, true)) throw new Exception('mkdir');
        }

        # WRITE
        $fh = fopen(DIR_LOG_ERROR . '/' . $logFile, "a");
        if (flock($fh, LOCK_EX)) {
            fwrite($fh, $logData);
            fflush($fh);
            flock($fh, LOCK_UN);
        }
        fclose($fh);
    }
} finally {
    header('Access-Control-Allow-Origin: *');
    header('Content-Type: text/plain');
    echo 'OK';
}


backend/cache.php
<?php
require_once dirname(__FILE__) . '/config.php';

# ================================================================================
# CONSTANTS

const PATH_ALL_PACKAGES = config\DIR_CACHE . '/all.json';
const URL_ALL_PACKAGES  = "https://package.elm-lang.org/all-packages";

/*
 * Version-tagged elm.json from GitHub should never change, so using longer expiry there.
 */
const REMOTE_LONG_EXPIRY_SECONDS        = 7 * 24 * 60 * 60;
const REMOTE_NORMAL_EXPIRY_SECONDS      = 1 * 24 * 60 * 60;
const REMOTE_NOTFOUND_EXPIRY_SECONDS    = 1 * 24 * 60 * 60;
const CLIENT_EXPIRY_SECONDS             = 1 * 24 * 60 * 60;

const JSON_SOFT_404                     = '{"error":"404"}';

const HTTP_STATUS_OK                    = 200;
const HTTP_STATUS_NOT_MODIFIED          = 304;
const HTTP_STATUS_NOT_FOUND             = 404;
const HTTP_STATUS_METHOD_NOT_ALLOWED    = 405;
const HTTP_STATUS_INTERNAL_SERVER_ERROR = 500;
const HTTP_STATUS_BAD_GATEWAY           = 502;

const EXAMPLES = [
    'example/via-core'         => '{"1.0.0":0,"1.0.1":0,"1.0.2":0}',
    'example/via-core/1.0.0'   => '{"dependencies":{"example/wants-core": "1.0.0 <= v < 1.0.1"}}',
    'example/via-core/1.0.1'   => '{"dependencies":{"example/wants-core": "1.0.1 <= v < 1.0.2"}}',
    'example/via-core/1.0.2'   => '{"dependencies":{"example/wants-core": "1.0.2 <= v < 1.0.3"}}',

    'example/via-http'         => '{"1.0.0":0,"2.0.0":0,"3.0.0":0}',
    'example/via-http/1.0.0'   => '{"dependencies":{"example/wants-http": "1.0.0 <= v < 2.0.0"}}',
    'example/via-http/2.0.0'   => '{"dependencies":{"example/wants-http": "2.0.0 <= v < 3.0.0"}}',
    'example/via-http/3.0.0'   => '{"dependencies":{"example/wants-http": "3.0.0 <= v < 4.0.0"}}',

    'example/wants-core'       => '{"1.0.0":0,"1.0.1":0,"1.0.2":0}',
    'example/wants-core/1.0.0' => '{"dependencies":{"elm/core": "1.0.0 <= v < 1.0.1"}}',
    'example/wants-core/1.0.1' => '{"dependencies":{"elm/core": "1.0.1 <= v < 1.0.2"}}',
    'example/wants-core/1.0.2' => '{"dependencies":{"elm/core": "1.0.2 <= v < 1.0.3"}}',

    'example/wants-http'       => '{"1.0.0":0,"2.0.0":0,"3.0.0":0}',
    'example/wants-http/1.0.0' => '{"dependencies":{"elm/http": "1.0.0 <= v < 2.0.0"}}',
    'example/wants-http/2.0.0' => '{"dependencies":{"elm/http": "2.0.0 <= v < 3.0.0"}}',
    'example/wants-http/3.0.0' => '{"dependencies":{"elm/http": "3.0.0 <= v < 4.0.0"}}',
];

# ================================================================================
# FUNCTIONS

/**
 * Fetch file from remote url and save to cache.
 *
 * Return simplified status code with cached data.
 *
 * Possible return values:
 * - [ HTTP_STATUS_OK         , $data ]
 * - [ HTTP_STATUS_NOT_FOUND  , null  ]
 * - [ HTTP_STATUS_BAD_GATEWAY, null  ]
 *
 */
# -------===========---------------------------
function FetchRemote($filePath, $url, $logId) {
# -------===========---------------------------
    $dirname      = dirname($filePath);
    $notFoundPath = $filePath . '.404';
    $etagPath     = $filePath . '.etag';
    $statusPath   = $filePath . '.status';
    $lockPath     = $filePath . '.fetching';

    $beginFloatTime = microtime(true);

    # MKDIR
    if (! file_exists($dirname)) {
        if (! mkdir($dirname, 0777, true)) throw new Exception("mkdir");
    }

    $timeBeforeLock = time();

    # ACQUIRE LOCK
#    RemoteLog(null, ">", null, $logId);
    $lock = fopen($lockPath, "c");
    if ($lock === false) throw new Exception("fopen - lockPath");
    flock($lock, LOCK_EX);

    try {
        # IF OTHER PROCESS JUST DID REMOTE FETCH, RETURN SAME THING
        if (file_exists($statusPath) && filemtime($statusPath) >= $timeBeforeLock) {
            $prevStatusStr = FileGetContentsOrThrow($statusPath, "statusPath");
            switch ($prevStatusStr) {
                case HTTP_STATUS_OK:
                    $data = FileGetContentsOrThrow($filePath, "filePath, re-OK");                    
                    $return = [ HTTP_STATUS_OK, $data ];
                    break;

                case HTTP_STATUS_NOT_FOUND:
                    $return = [ HTTP_STATUS_NOT_FOUND, null ];
                    break;

                case HTTP_STATUS_BAD_GATEWAY:
                    $return = [ HTTP_STATUS_BAD_GATEWAY, null ];
                    break;

                default:
                    throw new Exception('invalid previous status');
            }

            RemoteLog($beginFloatTime, "*", $return[0], $logId);
            return $return;
        }

        # OPTIONS
        $headers = [];
        $isConditionalRequest = false;
        if (file_exists($filePath) && file_exists($etagPath)) {
            $oldEtag = FileGetContentsOrThrow($etagPath, "etagPath");            

            if ($oldEtag !== "") {
                array_push($headers, "If-None-Match: \"$oldEtag\"");
                $isConditionalRequest = true;
            }
        }
        $options = [
            'http' => [
              'ignore_errors' => true,
              'header' => $headers,
            ]
        ];

        # FETCH
        $context = stream_context_create($options);
        $data    = file_get_contents($url, false, $context);
        list ($status, $etag) = ParseResponseHeaders($http_response_header);

        # HANDLE
        if ($status === HTTP_STATUS_OK) {
            if ($data === false) throw new Exception("file_get_contents ok, no data");

            # REMOVE 404
            if (file_exists($notFoundPath)) {
                if (! unlink($notFoundPath)) throw new Exception("unlink 404");
            }

            # SAVE TO CACHE
            $tmpPath = $filePath . '.tmp-' . microtime(true);
            try {
                if (file_put_contents($tmpPath, $data) !== strlen($data))
                    throw new Exception("file_put_contents - data");
                if (file_exists($filePath)) {
                    if (! unlink($filePath)) throw new Exception("unlink");
                }
                if (! rename($tmpPath, $filePath)) throw new Exception("rename");            
            } finally {
                @unlink($tmpPath);
            }

            # SAVE ETAG
            $etagOrEmpty = $etag ?? "";
            if (file_put_contents($etagPath, $etagOrEmpty) !== strlen($etagOrEmpty))
                throw new Exception("file_put_contents - etag");

            $return = [ HTTP_STATUS_OK, $data ];

        } elseif ($status === HTTP_STATUS_NOT_MODIFIED) {
            if (! touch($etagPath)) throw new Exception("touch etagPath");

            if (! file_exists($filePath)) throw new Exception("not modified -> ! file_exists");
            $data = FileGetContentsOrThrow($filePath, "filePath, NotModified");

            $return = [ HTTP_STATUS_OK, $data ];

        } elseif ($status === HTTP_STATUS_NOT_FOUND) {
            if (! touch($notFoundPath)) throw new Exception("touch notFoundPath");
            if (! touch($etagPath)) throw new Exception("touch etagPath");

            $return = [ HTTP_STATUS_NOT_FOUND, null ];

        } else {
            if (! touch($etagPath)) throw new Exception("touch etagPath");

            $return = [ HTTP_STATUS_BAD_GATEWAY, null ];
        }

        # SAVE STATUS
        if (file_put_contents($statusPath, $return[0]) !== strlen($return[0]))
            throw new Exception('file_put_contents - statusPath');

        RemoteLog($beginFloatTime, $isConditionalRequest ? "C" : "-", $status, $logId);

        # RETURN
        return $return;

    } finally {
#        RemoteLog(null, "<", null, $logId);
        flock($lock, LOCK_UN);
        fclose($lock);
        @unlink($lockPath);
    }
}

# -------======================-----------------------
function FileGetContentsOrThrow($path, $description) {
# -------======================-----------------------
    if (! file_exists($path)) throw new Exception("file_get_contents > file_exists: $description");
    $data = file_get_contents($path);
    if ($data === false) throw new Exception("file_get_contents: $description");
    return $data;
}

/**
 * Convert unixtime to format used e.g. by headers
 *   "Expires", "If-Modified-Since" and "Last-Modified"
 */
# -------========---------
function HttpDate($time) {
# -------========---------
    return gmdate("D, d M Y H:i:s", $time) . " GMT";
}

# -------====================------------
function ParseResponseHeaders($headers) {
# -------====================------------
    $status = null;
    $etag   = null;

    foreach ($headers as $header) {
        # In case of redirected request, there can be several HTTP status lines.
        if (preg_match('#^HTTP/\d\.\d\s(\d{3}) #', $header, $m) === 1) {
            $status = intval($m[1]);
        }
        if (preg_match('#^ETag:\s*(?:W\\\\)?"?(.+?)"?$#i', $header, $m) === 1) {
            $etag = $m[1];
        }
    }

    if ($status === null) throw new Exception('no status found in http_response_header');

    return [ $status, $etag ];
}


# -------=========-------------------------------------------
function RemoteLog($beginFloatTime, $type, $status, $logId) {
# -------=========-------------------------------------------
    list ($usecAsSec, $sec) = explode(" ", microtime());
    $usec    = intval($usecAsSec * 1000 * 1000);
    $logFile = config\DIR_LOG . '/' . gmdate("Y-m-d", $sec);
    $logLine = sprintf("%s.%06s  %5s  %5s %s  %s %s\n",
        gmdate("Y-m-d H:i:s", $sec),
        $usec,
        ($beginFloatTime == null ? "" : intval((microtime(true) - $beginFloatTime) * 1000)),
        getmypid(),
        $type,
        $status ?? "---",
        $logId
    );
    $fh = fopen($logFile, "a");
    if (flock($fh, LOCK_EX)) {
        fwrite($fh, $logLine);
        fflush($fh);
        flock($fh, LOCK_UN);
    }
    fclose($fh);
}

# -------=====--------------------------------------
function Serve($author, $project, $version = null) {
# -------=====--------------------------------------

    # ========================================
    # GET ALL-PACKAGES

    if (! file_exists(PATH_ALL_PACKAGES)
            || time() - filemtime(PATH_ALL_PACKAGES) > REMOTE_NORMAL_EXPIRY_SECONDS) {
        list ($allPackagesStatus, $allPackagesJson) =
            FetchRemote(PATH_ALL_PACKAGES, URL_ALL_PACKAGES, "ALL PACKAGES");
        if ($allPackagesJson === null) throw new Exception('FetchRemote - all packages');
    } else {
        $allPackagesJson = FileGetContentsOrThrow(PATH_ALL_PACKAGES, "all packages");
    }

    $allPackages = json_decode($allPackagesJson, true);
    if ($allPackages === null) throw new Exception("json_decode - all packages");

    # ========================================
    # VALIDATE REQUEST

    # VALIDATE AUTHOR/PROJECT
    if (! array_key_exists("$author/$project", $allPackages)) {
        ServePageNotFound(true);
        return;
    }

    # VALIDATE VERSION
    if ($version !== null && ! in_array($version, $allPackages["$author/$project"])) {
        ServePageNotFound(true);
        return;       
    }

    # ========================================
    # SETUP

    if ($version === null) {
        $logId        = "$author/$project";
        $filePath     = config\DIR_CACHE . "/$author/$project/releases.json";
        $remoteUrl    = "https://package.elm-lang.org/packages/$author/$project/releases.json";
        $remoteExpiry = REMOTE_NORMAL_EXPIRY_SECONDS;

    } else {
        $logId        = "$author/$project/$version";
        $filePath     = config\DIR_CACHE . "/$author/$project/$version.json";
        $remoteUrl    = "https://raw.githubusercontent.com/$author/$project/$version/elm.json";
        $remoteExpiry = REMOTE_LONG_EXPIRY_SECONDS;
    }

    $notFoundPath = $filePath . '.404';
    $etagPath     = $filePath . '.etag';

    # ========================================
    # SERVE

    if (file_exists($notFoundPath) && time() - filemtime($notFoundPath) < REMOTE_NOTFOUND_EXPIRY_SECONDS) {
        # SERVE CACHED NOT FOUND

        ServePageNotFound(true);
        return;

    } elseif (file_exists($filePath) && file_exists($etagPath)
            && time() - filemtime($etagPath) < $remoteExpiry) {
        # SERVE CACHED

        $data = FileGetContentsOrThrow($filePath, "filePath, cached");

        ServeJson($data, filemtime($filePath), true);
        return;

    } else {
        # FETCH FROM REMOTE SERVER

        list ($status, $data) = FetchRemote($filePath, $remoteUrl, $logId);

        if ($status === HTTP_STATUS_OK) {
            ServeJson($data, filemtime($filePath));
            return;

        } elseif ($status === HTTP_STATUS_NOT_FOUND) {
            ServePageNotFound(true);
            return;

        } else {
            ServeBadGateway();
            return;
        }
    } 
}

# -------===============----
function ServeBadGateway() {
# -------===============----
    header('Access-Control-Allow-Origin: *');
    http_response_code(HTTP_STATUS_BAD_GATEWAY);
}

# -------=========--------------------------------------------
function ServeJson($data, $mtime, $maybeNotModified = false) {
# -------=========--------------------------------------------

    $dateLastmod = HttpDate($mtime);
    $dateExpires = HttpDate(time() + CLIENT_EXPIRY_SECONDS);
    $etag        = md5($data);

    if ($maybeNotModified) {
        $ifModifiedSince = $_SERVER['HTTP_IF_MODIFIED_SINCE'] ?? false;
        $ifNoneMatch     = $_SERVER['HTTP_IF_NONE_MATCH'    ] ?? false;

        // If-Modified-Since is ignored if used with If-None-Match
        // > https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since

        if ($ifNoneMatch !== false) {
            // remove possible "W/" prefix
            if (substr($ifNoneMatch, 0, 2) === "W/") $ifNoneMatch = substr($ifNoneMatch, 0);

            // remove possible quotes (only " is actually allowed by specification)
            $ifNoneMatch = rtrim(ltrim($ifNoneMatch, "'\""), "'\"");

            if ($ifNoneMatch === $etag) {
                ServeNotModified($dateExpires, $etag);
                return;                
            }
        
        } elseif ($ifModifiedSince !== false && @strtotime($ifModifiedSince) >= $mtime) {
            ServeNotModified($dateExpires, $etag);
            return;
        }
    }

    header("Cache-Control: public, max-age=" . CLIENT_EXPIRY_SECONDS);
    header("Last-Modified: $dateLastmod");
    header("Expires: $dateExpires");
    header("ETag: \"$etag\"");

    header('Content-Type: application/json');
    header('Content-Length: ' . strlen($data));

    header('Access-Control-Allow-Origin: *');

    echo $data;
}

# -------=====================----
function ServeMethodNotAllowed() {
# -------=====================----
    http_response_code(HTTP_STATUS_METHOD_NOT_ALLOWED);
}

# -------================-----------------------
function ServeNotModified($dateExpires, $etag) {
# -------================-----------------------
    header("Cache-Control: public, max-age=" . CLIENT_EXPIRY_SECONDS);
    header("Expires: $dateExpires");
    header("ETag: \"$etag\"");

    header('Access-Control-Allow-Origin: *');
    http_response_code(HTTP_STATUS_NOT_MODIFIED);
}

# -------=================-----------------
function ServePageNotFound($soft = false) {
# -------=================-----------------
    if ($soft) {
        # soft errors are cached better than actual errors
        ServeJson(JSON_SOFT_404, time(), true);

    } else {
        header('Access-Control-Allow-Origin: *');
        http_response_code(HTTP_STATUS_NOT_FOUND);
    }
}

# ================================================================================
# MAIN

try {
    $method = $_SERVER['REQUEST_METHOD'];
    $query  = $_SERVER['QUERY_STRING'];

    if ($method !== "GET") {
        ServeMethodNotAllowed();
        exit;
    }

    # EXAMPLES
    if (array_key_exists($query, EXAMPLES)) {
        ServeJson(EXAMPLES[$query], time());
        exit;
    }

    # TODO: what characters are allowed for AUTHOR & PROJECT

    # "AUTHOR/PROJECT"
    if (preg_match("/^([0-9a-zA-Z-]+)\/([0-9a-zA-Z-]+)$/", $query, $m) === 1) {
        $author    = $m[1];
        $project   = $m[2];

        Serve($author, $project);
        exit;
    }

    # "AUTHOR/PROJECT/VERSION"
    if (preg_match("/^([0-9a-zA-Z-]+)\/([0-9a-zA-Z-]+)\/(\d+\.\d+\.\d+)$/", $query, $m) === 1) {
        $author    = $m[1];
        $project   = $m[2];
        $version   = $m[3];

        Serve($author, $project, $version);
        exit;
    }

    # SERVE NOT FOUND
    ServePageNotFound();
    exit;

} catch (Throwable $e) {
    header('Content-Type: text/plain');
    echo "ERROR";
    #echo $e->getMessage();
    http_response_code(HTTP_STATUS_INTERNAL_SERVER_ERROR);
}
